### PR TITLE
Extended multiblock rotation

### PR DESCRIPTION
**Pull request by serenibyss**
Freedom wrench but not a separate tool

I will move this branch to target 2.8 instead of master before merge

Currently tested and confirmed working:
- Multiblock forming code
- Structure preview in JEI
- Structure preview in-world
- Auto-build
- Part sorters (like DT and assembly line for specific multiblock part ordering) and relative facing api
- Rotation overlay on tool grid to indicate rotation when holding wrench
- Automatic structure check for the structure pattern flipped along the relative Y-axis. Allows mirroring the controller on multiblocks like the Assembly Line or Large Turbines

Multiblocks currently excluded from allowing extended rotation (may need to add more):
- Cleanroom
- Multiblock Miners/Fluid Drills
- Coke Oven + PBF (and by default disabled for any multi that implements the primitive multi base class)
- Primitive Pump
- Charcoal Pile Igniter
- Distillation Tower

### CODE REVIEW

**Comment by PrototypeTrousers** [comment]
1. WHY

**Comment by YoungOnionMC** [comment]
Shift right clicking with a wrench on the face should rotate the face counter clockwise 90*

**Review by ALongStringOfNumbers — COMMENTED** [review]
Should extended rotation also be disabled for the Steam Multiblocks, since it is disabled for the primitive ones?

Also, did we want to disable rotation on the central monitor, or leave it?

**Comment by Ghostipedia** [comment]
Day one of asking for PR review 🥺

**Review by ALongStringOfNumbers — APPROVED** [review]
Honestly, the Assembly Line particles not working for some rotated Assembly Lines is not blocking. But if possible I would like it resolved before the next release.

File: src/main/java/gregtech/api/metatileentity/multiblock/MultiblockControllerBase.java [hunk 1]
```diff
@@ -381,6 +482,35 @@ public boolean onRightClick(EntityPlayer playerIn, EnumHand hand, EnumFacing fac
...
+        }
+        return super.onWrenchClick(playerIn, hand, wrenchSide, hitResult);
+    }
+
+    @Override
+    public boolean isValidFrontFacing(EnumFacing facing) {
+        return allowsExtendedFacing() || super.isValidFrontFacing(facing);
+    }
+
+    // todo tooltip on multis saying if this is enabled or disabled?
```
Comment (TechLord22) [inline src/main/java/gregtech/api/metatileentity/multiblock/MultiblockControllerBase.java:503]: I think a tooltip for the disabled state is a good idea. Maybe put in the tool usages section?

File: src/main/java/gregtech/api/pattern/BlockPattern.java [hunk 2]
```diff
@@ -528,7 +541,7 @@ public BlockInfo[][][] getPreview(int[] repetition) {
                             }
                         }
                         BlockInfo info = infos == null || infos.length == 0 ? BlockInfo.EMPTY : infos[0];
-                        BlockPos pos = setActualRelativeOffset(z, y, x, EnumFacing.NORTH);
+                        BlockPos pos = setActualRelativeOffset(z, y, x, EnumFacing.NORTH, EnumFacing.UP, false);
```
Comment (TechLord22) [inline src/main/java/gregtech/api/pattern/BlockPattern.java:544]: Is always using `false` here intentional, or should it use `controllerBase.isFlipped()`?

Comment (serenibyss) [reply src/main/java/gregtech/api/pattern/BlockPattern.java:544]: Its intentional, its the JEI preview so it just forces to no rotation and no flip ("default state")

**Review by TechLord22 — APPROVED** [review]

**Pull request merged**